utilities/ggd/lib/go.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package lib
+
+import (
+	"path/filepath"
+	"strings"
+)
+
+// ChangedPackage is a changed package.
+type ChangedPackage struct {
+	Name        string
+	OnlyDeletes bool
+}
+
+// ChangedPackages returns all changed packages.
+func ChangedPackages(changedFiles []ChangedFile) []ChangedPackage {
+	changedPkgs := make(map[string]bool, len(changedFiles))
+	for _, f := range changedFiles {
+		if filterChanges(f.Filename) {
+			continue
+		}
+		dir := filepath.Dir(f.Filename)
+		if deletes, ok := changedPkgs[dir]; ok {
+			changedPkgs[dir] = deletes && f.Deleted
+		} else {
+			changedPkgs[dir] = f.Deleted
+		}
+	}
+	pkgs := make([]ChangedPackage, 0, len(changedPkgs))
+	for p, onlyDeletes := range changedPkgs {
+		pkgs = append(pkgs, ChangedPackage{
+			Name:        p,
+			OnlyDeletes: onlyDeletes,
+		})
+	}
+	return pkgs
+}
+
+// filterChanges filters changed files based on the following heuristic,
+// only files meeting any of the following conditions affect go compilation
+// 	- any .go file
+//  - any directory called `testdata/`
+//
+// also filters any `vendor/` directory paths
+func filterChanges(input string) bool {
+	if strings.Contains(input, "/vendor/") {
```

Comment:
Why filter the vendor? Can't we use the DAG to determine exactly what packages depend on them if we strip the vendor prefix?

Reply:
We can, will make it optional. 

I'm thinking of the case for the repos where we include vendor (internally) v the ones we don't (OSS/GH). To your point, it'll be helpful for the former, don't want to include it in the computation for the latter.

Reply:
done